src/serialize.ts

```diff
@@ -0,0 +1,133 @@
+    );
+  // the remaining chunks are stored as DataViews in msg.buffers
+  msg.buffers = [];
+  var start: number, stop: number;
+  for (i = 1; i < nbufs; i++) {
+    start = offsets[i];
+    stop = offsets[i + 1] || buf.byteLength;
+    msg.buffers.push(new DataView(buf.slice(start, stop)));
+  }
+  return msg;
+};
+
+
+/**
+ * deserialize the binary message format
+ * callback will be called with a message whose buffers attribute
+ * will be an array of DataViews.
+ */
+function _deserializeBinary(data: Blob | ArrayBuffer): IKernelMsg | Promise<IKernelMsg> {
+
+  if (data instanceof Blob) {
+    // data is Blob, have to deserialize from ArrayBuffer in reader callback
+    var reader = new FileReader();
+    var promise = new Promise(function(resolve, reject) {
+      reader.onload = function() {
+        var msg = _deserializeArrayBuffer((<ArrayBuffer>this.result));
+        resolve(msg);
+      };
+    });
+    reader.readAsArrayBuffer(data);
+    return promise;
+  } else {
+    // data is ArrayBuffer, can deserialize directly
+    var msg = _deserializeArrayBuffer((<ArrayBuffer>data));
+    return msg;
+  }
+};
+
+
+/**
+ * deserialize a message and return a promise for the unpacked message
+ */
+export
+function deserialize(data: Blob | ArrayBuffer | string): Promise<IKernelMsg> {
+  if (typeof data === "string") {
+    // text JSON message
+    return Promise.resolve(JSON.parse(data));
+  } else {
+    // binary message
+    return Promise.resolve(_deserializeBinary(data));
+  }
+};
+
+
+/**
+ * implement the binary serialization protocol
+ * serializes JSON message to ArrayBuffer
+ */
+function _serializeBinary(msg: IKernelMsg): ArrayBuffer {
+  var offsets: number[] = [];
+  var buffers: ArrayBuffer[] = [];
+  var i: number;
+  for (i = 0; i < msg.buffers.length; i++) {
+    // msg.buffers elements could be either views or ArrayBuffers
+    // buffers elements are ArrayBuffers
+    var b: any = msg.buffers[i];
+    buffers.push(b instanceof ArrayBuffer ? b : b.buffer);
+  }
+  delete msg.buffers;
+  var json_utf8 = (new TextEncoder('utf8')).encode(JSON.stringify(msg));
+  msg.buffers = buffers;
+  buffers.unshift(Array.prototype.slice.call(json_utf8));
+  var nbufs = buffers.length;
+  offsets.push(4 * (nbufs + 1));
+  for (i = 0; i + 1 < buffers.length; i++) {
+    offsets.push(offsets[offsets.length - 1] + buffers[i].byteLength);
+  }
+  var msg_buf = new Uint8Array(
+    offsets[offsets.length - 1] + buffers[buffers.length - 1].byteLength
+    );
+  // use DataView.setUint32 for network byte-order
+  var view = new DataView(msg_buf.buffer);
+  // write nbufs to first 4 bytes
+  view.setUint32(0, nbufs);
+  // write offsets to next 4 * nbufs bytes
+  for (i = 0; i < offsets.length; i++) {
+    view.setUint32(4 * (i + 1), offsets[i]);
+  }
+  // write all the buffers at their respective offsets
+  for (i = 0; i < buffers.length; i++) {
+    msg_buf.set(new Uint8Array(buffers[i]), offsets[i]);
+  }
+    
+  // return raw ArrayBuffer
```

Comment:
not sure how useful this comment is. Can probably delete it and the blank line above it.